### PR TITLE
move package loading behind a lock

### PR DESCRIPTION
**Pull request by KristofferC**
Ref https://github.com/JuliaIO/FileIO.jl/issues/336, not sure if it completely fixes the issue.

### CODE REVIEW

**Review by johnnychen94 — APPROVED** [review]
> Perhaps this indicates that Base.require itself is not threadsafe, because it appears to use plain Base.Condition, rather than Threads.Condition.

Just curious, could this be fixed in upstream Julia?

**Comment by KristofferC** [comment]
Maybe, but I don't think that lock is really for concurrency safety but to limit recursive `require` calls. But I am not sure.

**Comment by johnnychen94** [comment]
I'll set up the CI and check if it works; I have good reason to believe this is the right fix but we still need to appropriately test it.

**Comment by johnnychen94** [comment]
I've locally verified that this test 1c9db8f captures the #336, so I'll merge once CI passes.

File: src/loadsave.jl
Comment (johnnychen94) [inline src/loadsave.jl:195]: It looks like for Julia <= v1.3, `Threads.SpinLock()` is required here.

**Comment by johnnychen94** [comment]
Some strange behavior here when testing CSVFiles: `SystemError: dup: Bad file descriptor`. using `SpinLock` doesn't workaround it. Any ideas?

cc: @davidanthoff

**Comment by johnnychen94** [comment]
Given that multi-threads are not very well supported before Julia 1.3, I decided to only test this with Julia >= 1.3.

I've locally tested it for all minor versions from 1.3-1.6 on my macOS and Ubuntu. For Julia 1.1 it can be fixed by using `SpinLock`, but for Julia 1.0 and 1.2, it is still broken anyway.

This is only a partial fix to #336 for Julia >= 1.3.

**Comment by davidanthoff** [comment]
@johnnychen94 sorry, only catching up right now with notifications. Was the problem with CSVFiles.jl resolved before this was merged? Or is there still something I should look into?

**Comment by johnnychen94** [comment]
@davidanthoff It was then identified an unrelated issue. Sorry for the spam!

**Pull request merged**
